Fix for #896

embray opened the pull request:
This is my attempt to work around #896.  I've run an appveyor build on the final commit in this branch a few times now and had it pass consistently, though no 100% promises.

This mostly works by eliminating some code duplication by reusing the existing `wait_for_file` helper function, with a few small modifications to it.

Commit:
Attempt to fix failing tests on AppVeyor--reuse existing wait_for_file helper, and allow it to catch a broader range of OSErrors (which may also occur when, for example, attempting to delete the file) if it's still held open by either the test process or some background process).

Commit:
On Python 2 IOError is not a subclass of OSError, so catch both explicitly

Commit:
Double sleep time between attempts

Commit:
Allow wait_for_file to also wait on empty files.

Sleep between empty reads as well.

embray commented:
It looks like maybe the travis builds failed for unrelated reasons, but I could be wrong.

giampaolo commented on psutil/tests/__init__.py:
```diff
@@ -380,20 +375,26 @@ def wait_for_pid(pid, timeout=GLOBAL_TIMEOUT):
-            time.sleep(0.001)
+        except (IOError, OSError):
+            time.sleep(sleep_for)
+            sleep_for *= 2
```
This incremental timeout is a good change. I would put a limit though so that we sleep no more than, say, 0.01 secs per loop. That will make unit tests a bit faster.
`sleep_for *= 2` should become `sleep_for = min(sleep_for * 2, 0.01)`.
This way we'll wait:

```
0.001
0.002
0.004
0.008
0.01
0.01
...
```

embray replied:
Okay

giampaolo commented on psutil/tests/__init__.py:
```diff
@@ -380,20 +375,26 @@ def wait_for_pid(pid, timeout=GLOBAL_TIMEOUT):
             return data
-        except IOError:
-            time.sleep(0.001)
+        except (IOError, OSError):
```
This is too broad. We should only `pass` in case of `OSError` + `ENOENT` (meaning file is still not there) and in case of windows also take `WindowsError` + `errno == 32` (file used by another process) into account.
This way we'll get an exception if, e.g., the file cannot be opened or removed because of permission issues (we want to be aware of that).

embray replied:
Okay, that makes sense.

Commit:
Limit time to sleep between wait_for_file loops to 0.01 seconds so the tests don't wind up taking too long

Commit:
Narrower condition for retrying wait_for_file

Commit:
Right--keep forgetting IOError is not same as OSError on Python 2

embray commented:
Yeesh...I'll get this right eventually :)

Commit:
Further cleaned up the exception handling here to make sense on both
windows and non-windows, Python 2 and 3.

On Python 3 EnvironmentError is an alias for OSError.  On Python 2
WindowsError is a subclass of EnvironmentError.  So this will correctly
check the windows error, if it exists, in all cases.

On Travis we also need to catch ENXIO on Linux, and for some reason
EOPNOTSUPP on OSX.

embray commented:
Ok, Travis build now failing for a seemingly unrelated reason.

giampaolo commented on psutil/tests/__init__.py:
```diff
@@ -380,20 +379,36 @@ def wait_for_pid(pid, timeout=GLOBAL_TIMEOUT):
+            posix_errno = exc.errno
+            win_errno = getattr(exc, 'winerror', None)
+
+            if (posix_errno in (errno.ENOENT, errno.ENXIO, errno.EOPNOTSUPP) or
```
Why are ENXIO and EOPNOTSUPP necessary?

embray replied:
Have a look at https://travis-ci.org/giampaolo/psutil/builds/163674749

giampaolo replied:
What the...? I've never seen that! Thanks. Looks like this is good for merging.

giampaolo commented on psutil/tests/__init__.py:
```diff
@@ -62,6 +62,10 @@
 else:
     import imp as importlib
 
+if sys.platform.startswith('win'):
```
You can simply do `if WINDOWS:` instead

embray replied:
ah, nice

giampaolo commented:
https://travis-ci.org/giampaolo/psutil/jobs/163704213 looks like this broke unicode tests somehow

giampaolo commented:
Merged for now. I will try to fix the remaining issues as a second step.

embray commented:
I was seeing those failures in the last Travis build, but assumed it was unrelated to this change.  I could certainly believe it's possible they're related but I don't immediately see how.  Maybe you have some better ideas.

giampaolo commented:
I think I will refactor the whole thing in order to extrapolate the "wait" logic (sleep(), etc) which is currently shared in 4 different places. A `retry` decorator looks like a sensible choice to me. Incidentally I think that will mean removing most of your work/logic (sorry).

embray commented:
That's fine--it makes sense to extract out to a general retry method.   I had the same thought at the time but mostly wanted to focus for now on the exact area that was causing failures for continuous integration (since I want that to be fully working so I can move on to adding Cygwin builds as well).

giampaolo commented:
Yeah, hopefully we should stabilize all tests soon.

giampaolo commented:
Note: for https://travis-ci.org/giampaolo/psutil/jobs/164018394 I will provide a fix soon so don't mind about it.

giampaolo commented:
I added the retry decorator here: https://github.com/giampaolo/psutil/pull/901/files
Also, tests should be more stable now.
